aptos-move/framework/aptos-framework/sources/SharedAccount.move

```diff
@@ -0,0 +1,153 @@
+module AptosFramework::SharedAccount {
+    use Std::Errors;
+    use Std::Signer;
+    use Std::Vector;
+    use AptosFramework::Account;
+    use AptosFramework::Coin;
+    use AptosFramework::SimpleMap::{Self, SimpleMap};
+    use AptosFramework::ResourceAccount;
+
+    struct Shares has store {
+        numerator: u64,
+        denominator: u64,
+    }
+
+    // Resource representing a shared account
+    struct SharedAccount has key {
+        share_record: SimpleMap<address, Shares>,
+    }
+
+    const EINVALID_INIT_INPUT: u64 = 0;
+    const EACCOUNT_DNE: u64 = 1;
+    const EADDRESS_ALREADY_EXISTS: u64 = 2;
+    const EINVALID_NUMERATOR_DENOMINATOR_COMBINATIONS: u64 = 3;
+    const EINVALID_SIGNER: u64 = 4;
+    const EINSUFFICIENT_BALANCE: u64 = 5;
+
+    public(script) fun initialize_shared_account(source: &signer, seed: vector<u8>): (signer) {
+        assert!(Account::exists_at(Signer::address_of(source)), Errors::invalid_argument(EACCOUNT_DNE));
+        let resource_signer = ResourceAccount::create_resource_account(source, seed, Vector::empty());
+        if (!exists<SharedAccount>(Signer::address_of(&resource_signer))) {
+            move_to(
+                &resource_signer,
+                SharedAccount {
+                    share_record: SimpleMap::create(),
+                }
+            )
+        };
+        resource_signer
+    }
+
+    public(script) fun initialize(source: &signer, seed: vector<u8>, addresses: vector<address>, numerators: vector<u64>, denominator: u64): (signer) acquires SharedAccount {
+        assert!(Vector::length(&addresses) == Vector::length(&numerators), Errors::invalid_argument(EINVALID_INIT_INPUT));
+
+        let resource_signer = initialize_shared_account(source, seed);
+
+        let shared_account = borrow_global_mut<SharedAccount>(Signer::address_of(&resource_signer));
+
+        let sum: u64 = 0;
+        let i = 0;
+
+        while (i < Vector::length(&numerators)) {
+            let num = *Vector::borrow(&numerators, i);
+            let addr = *Vector::borrow(&addresses, i);
+            assert!(Account::exists_at(addr), Errors::invalid_argument(EACCOUNT_DNE));
+            sum = sum + num;
+            i = i + 1;
+
+            SimpleMap::add(&mut shared_account.share_record, addr, Shares { numerator: num, denominator: denominator });
```

Comment:
Do we need the total_shares in every mapping?  Or can we share it in the `SharedAccount`?

It's really low storage overhead, but if you have 100 shareholders, then you store it 100 times rather than once.

Reply:
ooh yea really good point! just realized that denominators could be different values so now we're passing in `denominators` as a vector